CONTRIBUTING_A_NEW_QUEST.md

```diff
@@ -0,0 +1,497 @@
+
+#### icon
+`override val icon = R.drawable.ic_quest_defibrillator` 
+icon drawable, you can initially use any icon. 
+
+Do not worry, submissions with placeholder icons are also welcomed! In many cases icon itself was not made by the PR author.
+
+More info about icon handling [will be given later](#adding-quest-icon).
+
+#### questTypeAchievements
+`override val questTypeAchievements = listOf(LIFESAVER)`
+
+In quest achievements, list what is relevant to the given quest, see the full list of available ones in [AchievementsModule.kt](app/src/main/java/de/westnordost/streetcomplete/data/user/achievements/AchievementsModule.kt)
+
+#### getTitle
+`override fun getTitle(tags: Map<String, String>) = R.string.quest_is_defibrillator_inside_title`
+
+It is a message displayed to user, code here passes a [reference](https://developer.android.com/guide/topics/resources/string-resource) to the string. You can change it to the new, not yet existing one and use a built in tool to place text.
+
+Actual strings sit in [app/src/main/res/values/strings.xml](app/src/main/res/values/strings.xml)
+
+There are separate files with translated text, but do not worry about it - [translations are handled separately](CONTRIBUTING.md#translating-the-app).
+
+#### Form
+
+`override fun createForm() = YesNoQuestAnswerFragment()`
+
+Form defines interface used by mappers.
+
+In this case, the simplest possible is used.
+
+But sometimes more complex ones are needed, see for example [AddBridgeStructure.kt](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/bridge_structure/AddBridgeStructure.kt) 
+
+`override fun createForm() = AddBridgeStructureForm()`
+
+
+With form defined in [AddBridgeStructureForm](app/src/main/java/de/westnordost/streetcomplete/quests/bridge_structure/AddBridgeStructureForm.kt)
+### applyAnswerTo
+```
+    override fun applyAnswerTo(answer: Boolean, tags: Tags, timestampEdited: Long) {
+        tags["indoor"] = answer.toYesNo()
+    }
+``` 
+
+This code is responsible for modifying `tags` object (passed by reference).
+
+In this case it would set `indoor` tag to have either `yes` or `no` answer, depending on selection in the quest interface.
+
+Actions may include (examples from various quests):
+
+- `tags["indoor"] = answer.toYesNo()`
+- `tags.remove("amenity")` - remove key if it is present
+- `tags.updateWithCheckDate("lit", answer.toYesNo())`
+  - in case of changing value such change will be applied
+  - in case of value being the same: add survey date tag, it would be `check_date:lit=` in this case
+     - so can be used even if value will stay the same - possible for resurvey quests
+     - survey date prefix is defined as [`const val SURVEY_MARK_KEY = "check_date"`](src/main/java/de/westnordost/streetcomplete/data/meta/OsmTaggings.kt)
+  - always update survey tag if present already
```

Comment:
```suggestion
  - `updateWithCheckDate()` method should always be used if your quest supports resurvey, in order for it to do resurveys at correct intervals (otherwise unrelated change to the node would reset resurvey counter)
  - if the new value differs from existing tag value: new value will be applied
  - if the new value is the same as existing tag value: adds [survey date tag](app/src/main/java/de/westnordost/streetcomplete/data/meta/OsmTaggings.kt#L33) (`check_date:lit=` in this example)
  - always update survey date tag (e.g.`check_date:lit=`) if it is present already
```

I think this is more clear and concise this way, while retaining all important information. line number should be kept up to date / forced to some commit version of file (same as all other links linking to line numbers elsewhere). 

While nice explanation and very useful for debug, I'm wondering if it would be more useful as a comment in the code around `updateWithCheckDate()` and thr "begginer quest writing docs" reduced just to `yout must use updateWithCheckDate() if y

Reply:
partially done, partially done differently, partially not fan of some ideas, partially migrated to https://github.com/matkoniecz/Zazolc/issues/9